src/pagination.rs

```diff
@@ -0,0 +1,279 @@
+use cobalt_model::pagination_config::PaginationCfg;
+use cobalt_model::permalink;
+use cobalt_model::SortOrder;
+use document;
+use document::Document;
+use liquid;
+use std::cmp::Ordering;
+use std::ops::Range;
+use std::path::PathBuf;
+
+pub(crate) fn generate_paginators(
+  doc: &mut Document,
+  posts_data: &[liquid::Value],
+  config: &mut PaginationCfg,
+) -> Vec<liquid::Object> {
+  let mut all_posts = posts_data.to_vec();
+  match config.include.as_str() {
+    "all" => {
+      sort_posts(&mut all_posts, &config);
+      create_all_paginators(&mut all_posts, &doc, &config)
+    }
+    _ => {
+      unreachable!();
+    }
+  }
+}
+
+fn extract_page_path_from(p: &liquid::Object) -> String {
+  p.get("page_path")
+    .expect("Should have a page_path")
+    .as_scalar()
+    .expect("Should be a scalar")
+    .to_str()
+    .into_owned()
+}
+
+pub(crate) fn extract_page_path(p: &liquid::Object, config: &PaginationCfg) -> PathBuf {
+  PathBuf::from(match config.include.as_str() {
+    "all" => extract_page_path_from(&p),
+    _ => "".to_owned(),
+  })
+}
+
+fn extract_value(a: &liquid::Value, key: &String) -> Option<liquid::Scalar> {
+  a.as_object().map_or(None, |attr| {
```

Comment:
Why aren't you using `Value::get`?

Reply:
Because I don't understand the method when I saw it: it returns an `Option<&Self>` and `Self` is a `Value`. But I want the `Scalar`.

Maybe I'm misunderstanding here, I'm still learning Rust so sorry for all these mistakes.

Reply:
Trying to use `get` like this:
```
a.get(key.clone()).map_or(None, |sort_key| {
    sort_key
      .as_scalar()
      .map_or(None, |value| Some(value.clone()))
  })
```

and got: 
```
error[E0277]: the trait bound `&liquid::Index: std::convert::From<std::string::String>` is not satisfied
  --> src\pagination.rs:45:5
   |
45 |   a.get(key.clone()).map_or(None, |sort_key| {
   |     ^^^ the trait `std::convert::From<std::string::String>` is not implemented for `&liquid::Index`
   |
   = help: the following implementations were found:
             <liquid::Index as std::convert::From<&'a str>>
             <liquid::Index as std::convert::From<std::string::String>>
   = note: required because of the requirements on the impl of `std::convert::Into<&liquid::Index>` for `std::string::String`
```
Which totally lost me.

Reply:
Maybe try
`a.get(&key.clone().into()).flat_map(|sort_key| sort_key.as_scalar())`

`Index` is a bit problematic in liquid's API.  At first, the error sounds weird but I think its complaining about only having ways to convert the param into an `Index` and not an `&Index`.

The last half I think is how to handle an `Option<Option<_>>` when working with `map`.

Reply:
I got:
```rust
error[E0599]: no method named `flat_map` found for type `std::option::Option<&liquid::Value>` in the current scope
  --> src\pagination.rs:54:30
   |
54 |   a.get(&key.clone().into()).flat_map(|sort_key| sort_key.as_scalar())
   |                              ^^^^^^^^
   |
   = note: the method `flat_map` exists but the following trait bounds were not satisfied:
           `&mut std::option::Option<&liquid::Value> : std::iter::Iterator`
```

Then I tried `a.get(&key.clone().into()).map_or(None, |k| k.flat_map(|sort_key| sort_key.as_scalar()))`
```rust
error[E0599]: no method named `flat_map` found for type `&liquid::Value` in the current scope
  --> src\pagination.rs:54:49
   |
54 |   a.get(&key.clone().into()).map_or(None, |k| k.flat_map(|sort_key| sort_key.as_scalar()))
   |                                                 ^^^^^^^^
   |
   = note: the method `flat_map` exists but the following trait bounds were not satisfied:
           `&mut &liquid::Value : std::iter::Iterator`
           `&mut liquid::Value : std::iter::Iterator`
```

Reply:
Cloning the key is an annoying part of liquid's API.

Do you need to clone the scalar?